### ISSUE TITLE
C# sometimes crashes during completion_queue_next on Mac

### ISSUE THREAD

**Issue by jtattermusch** [grpc/grpc#4968]
Occasionally I see a crash with this stacktrace:

```
	0   mono                                0x000d3186 mono_handle_native_sigsegv + 342
	1   mono                                0x00123891 sigabrt_signal_handler + 129
	2   libsystem_platform.dylib            0x90a2a03b _sigtramp + 43
	3   ???                                 0xffffffff 0x0 + 4294967295
	4   libsystem_c.dylib                   0x9a7d8dee abort + 156
	5   libgrpc_csharp_ext.dylib            0x0375b256 ref_by + 134
	6   libgrpc_csharp_ext.dylib            0x0375b4e7 grpc_fd_ref + 39
	7   libgrpc_csharp_ext.dylib            0x0375b8df grpc_fd_begin_poll + 63
	8   libgrpc_csharp_ext.dylib            0x03779d14 multipoll_with_poll_pollset_maybe_work_and_unlock + 900
	9   libgrpc_csharp_ext.dylib            0x0377ae8d grpc_pollset_work + 877
	10  libgrpc_csharp_ext.dylib            0x03750953 grpc_completion_queue_next + 1155
	11  libgrpc_csharp_ext.dylib            0x0373936e grpcsharp_completion_queue_next + 94
```

Some instances:
https://grpc-testing.appspot.com/job/gRPC_master/14743/config=dbg,language=csharp,platform=macos/testReport/junit/(root)/tests/csharp_Grpc_IntegrationTesting_InteropClientServerTest/

https://grpc-testing.appspot.com/job/gRPC_master/14748/config=dbg,language=csharp,platform=macos/testReport/junit/(root)/tests/csharp_Grpc_Core_Tests_MarshallingErrorsTest/

**Comment by jtattermusch** [comment]
@nicolasnoble    any wisdom?

**Comment by jtattermusch** [comment]
The root cause is:
`E1/29/2016 9:48:32 AM 2965557248 src/core/iomgr/fd_posix.c:126: assertion failed: gpr_atm_no_barrier_fetch_add(&fd->refst, n) > 0`

**Comment by jtattermusch** [comment]
seen again

https://grpc-testing.appspot.com/job/gRPC_master/15397/config=opt,language=csharp,platform=macos/testReport/junit/(root)/tests/csharp_Math_Tests_MathClientServerTest/

**Comment by ctiller** [comment]
Currently working on simplifying this code a lot. Once that's done, let's
nail this bug.

**Comment by jtattermusch** [comment]
Ah, great.

**Comment by jtattermusch** [comment]
Fixed by #5609.
